encoding/protobuf.go

```diff
@@ -0,0 +1,96 @@
+package encoding
+
+import (
+	"encoding/json"
+	"fmt"
+	"strings"
+
+	"github.com/golang/protobuf/proto"
+	"github.com/jhump/protoreflect/desc"
+	"github.com/jhump/protoreflect/dynamic"
+	"github.com/yarpc/yab/protobuf"
+	"github.com/yarpc/yab/transport"
+)
+
+type protoSerializer struct {
+	serviceName string
+	methodName  string
+	method      *desc.MethodDescriptor
+}
+
+// NewProtobuf returns a protobuf serializer.
+func NewProtobuf(fullMethodName string, source protobuf.DescriptorProvider) (Serializer, error) {
+	svc, mth, err := splitMethod(fullMethodName)
+	if err != nil {
+		return nil, err
+	}
+	dsc, err := source.FindSymbol(svc)
+	if err != nil {
+		return nil, fmt.Errorf("failed to query for service for symbol %q: %v", svc, err)
+	}
+	sd, ok := dsc.(*desc.ServiceDescriptor)
+	if !ok {
+		return nil, fmt.Errorf("target server does not expose service %q", svc)
+	}
+	mtd := sd.FindMethodByName(mth)
+	if mtd == nil {
+		return nil, fmt.Errorf("service %q does not include a method named %q", svc, mth)
+	}
+	return &protoSerializer{
+		serviceName: svc,
+		methodName:  mth,
+		method:      mtd,
+	}, nil
+}
+
+func (p protoSerializer) Encoding() Encoding {
+	return Protobuf
+}
+
+func (p protoSerializer) Request(body []byte) (*transport.Request, error) {
+	req := dynamic.NewMessage(p.method.GetInputType())
+	if err := req.UnmarshalJSON(body); err != nil {
+		return nil, fmt.Errorf("could not parse given request body as message of type %q: %v", p.method.GetInputType().GetFullyQualifiedName(), err)
+	}
+	bytes, err := proto.Marshal(req)
+	if err != nil {
+		return nil, fmt.Errorf("could marshal message of type %q: %v", p.method.GetInputType().GetFullyQualifiedName(), err)
+	}
+	return &transport.Request{
+		Method: fmt.Sprintf("%s::%s", p.serviceName, p.methodName),
```

Comment:
This is fine, but string concatenation would work, `p.serviceName + "::" + p.methodName`, and avoids the surprising cost of `...interface {}`.

Reply:
Just for my curiousity -- is this actually used by protobuf or only YARPC?

Reply:
Currently, this is a yarpc thing. The outbound request is made by the yarpc library and splits the procedure on `::` and to convert it to the grpc wire format.